packages/go/analysis/ad/adcs.go

```diff
@@ -161,6 +174,157 @@ func PostADCSESC3(ctx context.Context, tx graph.Transaction, outC chan<- analysi
 	return nil
 }
 
+func principalControlsCertTemplate(principal, certTemplate *graph.Node, groupExpansions impact.PathAggregator, cache ADCSCache) bool {
+	var (
+		expandedTemplateControllers = cache.ExpandedCertTemplateControllers[certTemplate.ID]
+		principalID                 = principal.ID.Uint32()
+	)
+
+	if slices.Contains(expandedTemplateControllers, principalID) {
+		return true
+	}
+
+	if CalculateCrossProductNodeSets(groupExpansions, graph.NewNodeSet(principal).Slice(), cache.CertTemplateControllers[certTemplate.ID]).Contains(principalID) {
+		cache.ExpandedCertTemplateControllers[certTemplate.ID] = append(expandedTemplateControllers, principalID)
+		return true
+	}
+
+	return false
+}
+
+func checkEmailValidity(node *graph.Node, validCertTemplates []*graph.Node, groupExpansions impact.PathAggregator, cache ADCSCache) bool {
+	exception := false
+	email, err := node.Properties.Get(common.Email.String()).String()
+	if err != nil {
+		log.Errorf("%s property access error %d: %v", common.Email.String(), node.ID, err)
+	}
+
+	if email == "" {
+		for _, certTemplate := range validCertTemplates {
+			if principalControlsCertTemplate(node, certTemplate, groupExpansions, cache) {
+				if schemaVersion, err := certTemplate.Properties.Get(ad.SchemaVersion.String()).Float64(); err != nil {
+					log.Errorf("%s property access error %d: %v", ad.SchemaVersion.String(), certTemplate.ID, err)
+					continue
+				} else if subjectAltRequireEmail, err := certTemplate.Properties.Get(ad.SubjectAltRequireEmail.String()).Bool(); err != nil {
+					log.Errorf("%s property access error %d: %v", ad.SubjectAltRequireEmail.String(), certTemplate.ID, err)
+					continue
+				} else if subjectRequireEmail, err := certTemplate.Properties.Get(ad.SubjectRequireEmail.String()).Bool(); err != nil {
+					log.Errorf("%s property access error %d: %v", ad.SubjectRequireEmail.String(), certTemplate.ID, err)
+					continue
+				} else if subjectAltRequireDNS, err := certTemplate.Properties.Get(ad.SubjectAltRequireDNS.String()).Bool(); err != nil {
+					log.Errorf("%s property access error %d: %v", ad.SubjectAltRequireDNS.String(), certTemplate.ID, err)
+					continue
+				} else if subjectAltRequireDomainDNS, err := certTemplate.Properties.Get(ad.SubjectAltRequireDomainDNS.String()).Bool(); err != nil {
+					log.Errorf("%s property access error %d: %v", ad.SubjectAltRequireDomainDNS.String(), certTemplate.ID, err)
+					continue
+				} else if node.Kinds.ContainsOneOf(ad.User) && (subjectAltRequireDNS || subjectAltRequireDomainDNS) {
+					continue
+				} else if (!subjectAltRequireEmail && !subjectRequireEmail) || schemaVersion == 1 {
+					exception = true
```

Comment:
I believe you can remove the exception var and return true here instead

Reply:
You're right! There wouldn't be a need to check any other certTemplates so an early return here is good. I've made this change.